library/x509_crt.c

```diff
@@ -1473,6 +1792,56 @@ static int x509_info_subject_alt_name( char **buf, size_t *size,
     return( 0 );
 }
 
+int mbedtls_x509_parse_subject_alt_name( const mbedtls_x509_buf *san_buf,
+                                         mbedtls_x509_subject_alternative_name *san )
+{
+    int ret;
+    switch( san_buf->tag &
+            ( MBEDTLS_ASN1_TAG_CLASS_MASK |
+              MBEDTLS_ASN1_TAG_VALUE_MASK ) )
+    {
+        /*
+         * otherName
+         */
+        case( MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_X509_SAN_OTHER_NAME ):
+        {
+            mbedtls_x509_san_other_name other_name;
+
+            ret = x509_get_other_name( san_buf, &other_name );
+            if( ret != 0 )
+                return( ret );
+
+            memset( san, 0, sizeof( mbedtls_x509_subject_alternative_name ) );
+            san->type = MBEDTLS_X509_SAN_OTHER_NAME;
+            memcpy( &san->san.other_name,
```

Comment:
Blocker: This structure contains pointers to an internal structure with an independent lifetime. This shallow copy hands over these pointers to the caller.

Reply:
what pointers? It only contains the `buf.p` pointer, which point to the raw data of the certificate. This is same as previous implementation.
There is no `next` pointer anymore

Reply:
Yes, it points to the raw data of the certificate and it is not good:
- If the certificate is freed and maybe even reallocated before the user processes the name, it will interpret data that is found at the address at that time as the name.
- If the user wants avoid memory leaks and frees the pointer. They will call free on the middle of the buffer. What makes it worse that address might have been freed or even reallocated by that time.

We can mitigate this by either:
1. Allocating a new buffer and doing a deep copy of the structure.
2. Documenting it clearly and in detail to avoid these mistakes.

This behaviour is really against encapsulation and programmer intuition, therefore I would advise to do 1. unless Hanno's in-place parsing PR does 2.

Reply:
I share @yanesca's concerns, but it should be noted that we already have this behaviour in the public API `mbedtls_asn1_get_sequence_of()`. There is no precedent for this in the existing X.509 module since, as far as I see,  everything that can be parsed is parsed and stored in `mbedtls_x509_crt`.

For on-demand parsing, it removes the heap-allocated linked lists (subject, issuer, SAN, ExtKeyUsage) from the CRT structure and offers to generate them through new convenience APIs. The implementation of this convenience API currently also uses shallow copies and is hence subject to the same concerns. I have no strong preference as to whether they should make a deep-copy, apart from the fact that it would also require a new kind of freeing function which not only frees the constituents of the linked list, but also the buffers referenced in each item. This increases complexity and code-size, but given that using the linked lists is discouraged on constrained systems anyway, it's not a blocker for me.

However, the situation is different here: We want to provide additional parsing functionality, and the approach taken in this PR is to not add that by default to `mbedtls_x509_crt` but to allow it to be requested if needed - which is good, because it avoids memory and code-size overhead for users that don't need it. But, in the interest of those users that _do_ need the new feature _and_ are memory conscious, the new functionality should be made available in a way which doesn't involve heap allocations and copies, and I think the way @RonEld does it here it the right way.

That doesn't take away the danger that @yanesca points out, but it leads me to strongly prefer (2) over (1): We should clearly document (both here and - that's my job - in the on-demand parsing PR) that the lifetime of the structures is restricted by the lifetime of the CRT.

Reply:
If we do it in the on-demand parsing PR too, that makes the situation much better: user mistakes are less likely, because they are more likely to expect this behaviour and consult the documentation if in doubt.